File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    int d = 0;
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        if (root->left == nullptr && root->right == nullptr) {
            return 0;
        }
        int leftHeight = (root->left != nullptr ? 1 : 0) + binaryTreeHeight(root->left);
        int rightHeight = (root->right != nullptr ? 1 : 0) + binaryTreeHeight(root->right);
        return max(d, leftHeight + rightHeight);
    }

    int binaryTreeHeight(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        if (root->left == nullptr && root->right == nullptr) {
            return 0;
        }

        int leftHeightTrimmed = binaryTreeHeight(root->left);
        int rightHeightTrimmed = binaryTreeHeight(root->right);

        d = max(d, leftHeightTrimmed + rightHeightTrimmed + (root->left != nullptr ? 1 : 0) + (root->right != nullptr ? 1 : 0));

        return max(1 + leftHeightTrimmed, 1 + rightHeightTrimmed);
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def visitNode(self, node: TreeNode, maxBefore: int):
        if not node:
            return
        if maxBefore <= node.val:
            self.goodNodesNum += 1
            maxBefore = node.val
        self.visitNode(node.left, maxBefore)
        self.visitNode(node.right, maxBefore)
    def goodNodes(self, root: TreeNode) -> int:
        self.goodNodesNum = 0
        self.visitNode(root, root.val)
        return self.goodNodesNum